### PR TITLE
ci: Add native macOS arm64 job

### PR DESCRIPTION
**Pull request by hebasto**
This PR starts using the [new](https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/) M1 macOS runner.

The alternative approach might be using a matrix, but it is not trivial to implement.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -585,10 +585,10 @@ jobs:
         run: env
         if: ${{ always() }}
 
-  macos-native:
-    name: "x86_64: macOS Monterey"
+  x86_64-macos-native:
+    name: "x86_64: macOS Monterey, Valgrind"
```
Comment (real-or-random) [inline .github/workflows/ci.yml:589]: Did you remove the "Valgrind" intentionally?

**Review by real-or-random — COMMENTED** [review]
Concept ACK

It's a bit sad that we can't run Valgrind on the new machines yet.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (hebasto) [reply .github/workflows/ci.yml:589]: Do you mean in the https://github.com/bitcoin-core/secp256k1/pull/1412?

Comment (real-or-random) [reply .github/workflows/ci.yml:589]: No, I mean you have removed the substring ", Valgrind" from the job name. :D

Comment (hebasto) [reply .github/workflows/ci.yml:589]: Sorry, I don't follow... The diff _appends_ the substring ", Valgrind" to the job name, no?

Comment (real-or-random) [reply .github/workflows/ci.yml:589]: Lol sure, sorry, nevermind...

**Comment by real-or-random** [comment]
I think it will be good to add an MSan job with ctimetests enabled. We have an arm64 job with Valgrind on Linux (Cirrus CI), but I guess having a ctimetest for macOS (Apple clang) can't hurt.

**Comment by hebasto** [comment]
> I think it will be good to add an MSan job with ctimetests enabled. We have an arm64 job with Valgrind on Linux (Cirrus CI), but I guess having a ctimetest for macOS (Apple clang) can't hurt.

For both Apple Clang and LLVM Clang, I have:
```
clang: error: unsupported option '-fsanitize=memory' for target 'arm64-apple-darwin23.5.0'
```

**Comment by real-or-random** [comment]
Ok, damn, I didn't know that MSan is not supported on macOS... https://github.com/google/sanitizers/issues/588

**Review by real-or-random — APPROVED** [review]
ACK 218f0cc93bd77eae926f4c4e259951b92bf00aec

**Pull request merged**